File: my-solutions/0105-construct-binary-tree-from-preorder-and-inorder-traversal/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* build(vector<int>& preorder, vector<int>& inorder, int& index, int i, int j) {
        if(i > j) {
            return nullptr;
        }

        TreeNode* root = new TreeNode(preorder[index]);
        int split = 0;
        for(int i = 0; i < size(inorder); i++) {
            if(inorder[i] == preorder[index]) {
                split = i;
            }
        }
        index++;
        root->left = build(preorder, inorder, index, i, split - 1);
        root->right = build(preorder, inorder, index, split + 1, j);

        return root;
    }
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        int index = 0;
        TreeNode* root = build(preorder, inorder, index, 0, size(inorder) - 1);
        return root;
    }
};

File: my-solutions/0105-construct-binary-tree-from-preorder-and-inorder-traversal/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        pre = 0
        def build(ino_l, ino_r):
            nonlocal pre
            if ino_l > ino_r:
                return None

            root = TreeNode(preorder[pre])
            idx = inorder.index(preorder[pre])
            pre += 1
            left = build(ino_l, idx - 1)
            right = build(idx + 1, ino_r)
            root.left = left
            root.right = right
            return root

        return build(0, len(inorder) - 1)

